### PR TITLE
FEATURE: Alert on mention that allows inviting users to join channel

### PR DESCRIPTION
**Pull request by markvanlan**
Relies on core PR https://github.com/discourse/discourse/pull/15288

Screen Shot 2021-12-13 at 10 56 08 PM

Screen Shot 2021-12-13 at 10 56 16 PM

Screen Shot 2021-12-13 at 10 56 44 PM

### CODE REVIEW

**Comment by SamSaffron** [comment]
for tomorrow, we have a failing spec.

File: assets/javascripts/discourse/components/tc-message.js [hunk 1]
```diff
@@ -66,6 +67,10 @@ export default Component.extend({
       this,
       "_handleReactionMessage"
     );
+
+    if (this._invitationSentTimer) {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/components/tc-message.js:71]: you can just do `cancel(foo)` if foo is `null` that will do nothing.

File: assets/javascripts/discourse/components/tc-message.js [hunk 2]
```diff
@@ -228,6 +233,43 @@ export default Component.extend({
     return Object.values(reactions).some((r) => r.count > 0);
   },
 
+  @discourseComputed("message.mentionWarning.cannot_see")
+  mentionedCannotSeeText(users) {
+    return I18n.t("chat.mention_warning.cannot_see", {
+      usernames: users.map((u) => u.username).join(", "),
```
Comment (jjaffeux) [inline assets/javascripts/discourse/components/tc-message.js:239]:
```suggestion
      usernames: users.mapBy("username").join(", "),
```

File: assets/javascripts/discourse/components/tc-message.js [hunk 3]
```diff
@@ -228,6 +233,43 @@ export default Component.extend({
...
+      count: users.length,
+    });
+  },
+
+  @discourseComputed("message.mentionWarning.without_membership")
+  mentionedWithoutMembershipText(users) {
+    return I18n.t("chat.mention_warning.without_membership", {
+      usernames: users.map((u) => u.username).join(", "),
```
Comment (jjaffeux) [inline assets/javascripts/discourse/components/tc-message.js:247]:
```suggestion
      usernames: users.mapBy("username").join(", "),
```

File: assets/javascripts/discourse/components/tc-message.js [hunk 4]
```diff
@@ -228,6 +233,43 @@ export default Component.extend({
...
+      count: users.length,
+    });
+  },
+
+  @action
+  inviteMentioned() {
+    const user_ids = this.message.mentionWarning.without_membership.map(
+      (u) => u.id
+    );
```
Comment (jjaffeux) [inline assets/javascripts/discourse/components/tc-message.js:256]:
```suggestion
    const user_ids = this.message.mentionWarning.without_membership.mapBy("id");
```

File: assets/javascripts/discourse/components/chat-live-pane.js [hunk 5]
```diff
@@ -586,6 +589,13 @@ export default Component.extend({
     return 0;
   },
 
+  handleMentionWarning(data) {
+    const message = this.messageLookup[data.chat_message_id];
+    if (message) {
+      message.set("mentionWarning", data);
+    }
```
Comment (jjaffeux) [inline assets/javascripts/discourse/components/chat-live-pane.js:596]: this.messageLookup[data.chat_message_id]?.set("mentionWarning", data);

Comment (jjaffeux) [reply assets/javascripts/discourse/components/chat-live-pane.js:596]: This should work I think

File: assets/javascripts/discourse/templates/components/tc-message.hbs [hunk 6]
```diff
@@ -204,6 +204,36 @@
...
+                }}
+
+                {{#if message.mentionWarning.cannot_see}}
+                  <p class="cannot-see">{{mentionedCannotSeeText}}</p>
+                {{/if}}
+
+                {{#if message.mentionWarning.without_membership}}
+                  <p class="without-membership">
+                    {{mentionedWithoutMembershipText}}
+                    <span class="invite-link" role="button" onclick={{action "inviteMentioned"}}>
```
Comment (jjaffeux) [inline assets/javascripts/discourse/templates/components/tc-message.hbs:228]: why not use a button directly ?

File: assets/javascripts/discourse/templates/components/tc-message.hbs [hunk 7]
```diff
@@ -204,6 +204,36 @@
               {{i18n "chat.send_error"}}
             </div>
           {{/if}}
+
+          {{#if message.mentionWarning}}
+            <div class="alert alert-info chat-message-mention-warning">
+              {{#if message.mentionWarning.invitationSent}}
+                {{d-icon "check"}}
+                {{i18n "chat.mention_warning.invitations_sent" count=message.mentionWarning.without_membership.length}}
```
Comment (jjaffeux) [inline assets/javascripts/discourse/templates/components/tc-message.hbs:212]:
```suggestion
                <span>{{i18n "chat.mention_warning.invitations_sent" count=message.mentionWarning.without_membership.length}}</span>
```

File: assets/javascripts/discourse/templates/components/tc-message.hbs [hunk 8]
```diff
@@ -204,6 +204,36 @@
...
+
+                {{#if message.mentionWarning.cannot_see}}
+                  <p class="cannot-see">{{mentionedCannotSeeText}}</p>
+                {{/if}}
+
+                {{#if message.mentionWarning.without_membership}}
+                  <p class="without-membership">
+                    {{mentionedWithoutMembershipText}}
+                    <span class="invite-link" role="button" onclick={{action "inviteMentioned"}}>
+                      {{i18n "chat.mention_warning.invite"}}
```
Comment (jjaffeux) [inline assets/javascripts/discourse/templates/components/tc-message.hbs:229]:
```suggestion
                      <span>{{i18n "chat.mention_warning.invite"}}</span>
```

File: assets/javascripts/discourse/widgets/chat-invitation-notification-item.js [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    const notificationName = lookup[notificationType];
+    const { data } = attrs;
+    const text = this.text(data);
+    const title = this.notificationTitle(notificationName, data);
+    const html = new RawHtml({ html: `<div>${text}</div>` });
+    const contents = [iconNode("link"), html];
+    return h("a", { attributes: { title } }, contents);
+  },
+
+  click() {
```
Comment (jjaffeux) [inline assets/javascripts/discourse/widgets/chat-invitation-notification-item.js:31]: should we keep the event from bubbling after or is this already handled? (eg: does the parent node receives the click event?)

File: lib/chat_notifier.rb [hunk 10]
```diff
@@ -109,7 +121,31 @@ def users_here(mention_matches)
...
+      if user.user_chat_channel_memberships.detect { |m| m.chat_channel_id == @chat_channel.id && m.following == true }
+        with_membership << user
+      else
+        without_membership << user
+      end
+    end
+    [with_membership, without_membership]
+  end
+
+  def mentioned_by_username(exclude:, usernames: nil)
```
Comment (CvX) [inline lib/chat_notifier.rb:137]: Is it ever called without `usernames:` arg?

File: lib/chat_notifier.rb [hunk 11]
```diff
@@ -123,10 +159,17 @@ def users_for_channel(exclude:, usernames: nil)
...
+    users.each do |user|
       guardian = Guardian.new(user)
-      guardian.can_chat?(user) && guardian.can_see_chat_channel?(@chat_channel)
+      if guardian.can_chat?(user) && guardian.can_see_chat_channel?(@chat_channel)
+        can_chat << user
+      else
+        cannot_chat << user
+      end
     end
+    [can_chat, cannot_chat]
```
Comment (CvX) [inline lib/chat_notifier.rb:172]: Just an FYI, there's a neat method just for this:

```ruby
users.partition do |user|
  guardian = Guardian.new(user)
  guardian.can_chat?(user) && guardian.can_see_chat_channel?(@chat_channel)
end
```

Comment (jjaffeux) [reply lib/chat_notifier.rb]: I love when we get opportunity to use this kind of methods :D

Comment (markvanlan) [reply lib/chat_notifier.rb]: Can't believe I didn't know this existed. So freaking awesome.

**Review by CvX — APPROVED** [review]

**Pull request merged**